scipy/sparse/_compressed.py

```diff
@@ -366,105 +362,127 @@ def multiply(self, other):
+                # build diagonal csc_array/csr_array => self._csr_container
+                data = other.toarray().ravel()
+                indptr = np.arange(oN + 1)
+                indices = indptr[:-1]
+                new_other = self._csr_container((data, indices, indptr))
+                result = self._matmul_sparse(new_other)
+                return result if self.ndim == 2 else result.reshape((1, oN))
             # self is a row.
-            elif self.shape[0] == 1 and self.shape[1] == other.shape[1]:
-                copy = self._dia_container(
-                    (self.toarray().ravel(), [0]),
-                    shape=(self.shape[1], self.shape[1])
-                )
+            if sM == 1 and sN == oN:
+                data = self.toarray().ravel()
+                indptr = np.arange(oN + 1)
+                indices = indptr[:-1]
+                copy = self._csr_container((data, indices, indptr))
                 return other._matmul_sparse(copy)
-            # Column vector times matrix. other is a column.
-            elif other.shape[1] == 1 and self.shape[0] == other.shape[0]:
-                other = self._dia_container(
-                    (other.toarray().ravel(), [0]),
-                    shape=(other.shape[0], other.shape[0])
-                )
-                return other._matmul_sparse(self)
+
+            # Other is a column.
+            if oN == 1 and sM == oM:
+                data = other.toarray().ravel()
+                indptr = np.arange(oM + 1)
+                indices = indptr[:-1]
+                new_other = self._csr_container((data, indices, indptr))
+                return new_other._matmul_sparse(self)
             # self is a column.
-            elif self.shape[1] == 1 and self.shape[0] == other.shape[0]:
-                copy = self._dia_container(
-                    (self.toarray().ravel(), [0]),
-                    shape=(self.shape[0], self.shape[0])
-                )
-                return copy._matmul_sparse(other)
-            else:
-                raise ValueError("inconsistent shapes")
+            if sN == 1 and sM == oM:
+                data = self.toarray().ravel()
+                indptr = np.arange(oM + 1)
+                indices = indptr[:-1]
+                new_self = self._csr_container((data, indices, indptr))
+                return new_self._matmul_sparse(other)
+            raise ValueError("inconsistent shapes")
 
         # Assume other is a dense matrix/array, which produces a single-item
         # object array if other isn't convertible to ndarray.
-        other = np.atleast_2d(other)
+        other = np.asanyarray(other)
 
-        if other.ndim != 2:
+        if other.ndim > 2:
             return np.multiply(self.toarray(), other)
         # Single element / wrapped object.
         if other.size == 1:
-            if other.dtype == np.object_:
-                # 'other' not convertible to ndarray.
-                return NotImplemented
```

Comment:
I think we still want to keep this check, because this is what allows us to handle types that override `__mul__` but don't convert via the `asanyarray()` call above.